vtysh: fix duplicated output of key chain configuration

idryzhov opened the pull request:
When both ripd and eigrpd run at the same time, all key configuration in
key chain node is duplicated. This change adds a concept of nested nodes
into vtysh to fix the issue.

Before:
```
key chain test
 key 1
  key-string 1
 exit
 key 1
  key-string 1
 exit
!
```
After:
```
key chain test
 key 1
  key-string 1
 exit
!
```

Signed-off-by: Igor Ryzhov <iryzhov@nfware.com>

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/e596bbe4198d98fec22ced9e849d3718/raw/68f3211e880a324f312abddfd9239513e508b753/cr_9396_1628896431.diff | git apply
```

</p>

<p>

```diff
diff --git a/vtysh/vtysh_config.c b/vtysh/vtysh_config.c
index 226041bf70..8790027773 100644
--- a/vtysh/vtysh_config.c
+++ b/vtysh/vtysh_config.c
@@ -280,9 +280,8 @@ void vtysh_config_parse_line(void *arg, const char *line)
 		if (config) {
 			if (config->index == KEYCHAIN_NODE
 			    && strncmp(line, " key", strlen(" key")) == 0) {
-				config = config_get_nested(config,
-							   KEYCHAIN_KEY_NODE,
-							   line);
+				config = config_get_nested(
+					config, KEYCHAIN_KEY_NODE, line);
 			} else if (config->index == KEYCHAIN_KEY_NODE) {
 				if (strncmp(line, " exit", strlen(" exit"))
 				    == 0) {
@@ -294,7 +293,8 @@ void vtysh_config_parse_line(void *arg, const char *line)
 							     line);
 				}
 			} else if (strncmp(line, " link-params",
-					   strlen(" link-params")) == 0) {
+					   strlen(" link-params"))
+				   == 0) {
 				config_add_line(config->line, line);
 				config->index = LINK_PARAMS_NODE;
 			} else if (strncmp(line, " ip multicast boundary",
@@ -302,7 +302,8 @@ void vtysh_config_parse_line(void *arg, const char *line)
 				   == 0) {
 				config_add_line_uniq_end(config->line, line);
 			} else if (strncmp(line, " ip igmp query-interval",
-					   strlen(" ip igmp query-interval")) == 0) {
+					   strlen(" ip igmp query-interval"))
+				   == 0) {
 				config_add_line_uniq_end(config->line, line);
 			} else if (config->index == LINK_PARAMS_NODE
 				   && strncmp(line, "  exit-link-params",
@@ -319,7 +320,8 @@ void vtysh_config_parse_line(void *arg, const char *line)
 				   || !strncmp(line, " no vrrp",
 					       strlen(" no vrrp"))) {
 				config_add_line(config->line, line);
-			} else if (!strncmp(line, " ip mroute", strlen(" ip mroute"))) {
+			} else if (!strncmp(line, " ip mroute",
+					    strlen(" ip mroute"))) {
 				config_add_line_uniq_end(config->line, line);
 			} else if (config->index == RMAP_NODE
 				   || config->index == INTERFACE_NODE

```

</p>
</details>

            
If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

rameshabhinay commented:
@idryzhov What is your opinion ? which solution do you prefer ? #8890 or #9396 ?

idryzhov commented:
@rameshabhinay you probably wanted to mention someone else, because this one is mine :)

rameshabhinay commented:
@idryzhov Its fine. If you want to push your changes. I am fine with it. I don't have any problems.
You can choose the best. 
Thanks

pushpasis commented on vtysh/vtysh_config.c:
```diff
@@ -104,22 +106,22 @@ struct configuration {
 };
 
-static struct config *config_get(int index, const char *line)
+static struct config *__config_get(vector vec, int index, const char *line)
```
Not sure if this is a good practice or not as per the frr coding guidelines to start the function name with '__'. Suggest to rename this to something without '__' at the beginning.

pushpasis replied:
Maybe rename this to 'config_gett_from_vec()'. It seems to be what its really doing.

pushpasis reviewed (approved):
@idryzhov Just have two nits for you. It would be great if you address it. Rest of the code looks quiet simple enough and good.

riw777 reviewed (approved):
waiting on other reviews, but looks good at this point

idryzhov commented:
ci:rerun

qlyoung commented on vtysh/vtysh_config.c:
```diff
@@ -507,23 +540,40 @@ void vtysh_config_dump(void)
 		}
+}
+
+/* Display configuration to file pointer. */
```
Not seeing the file pointer in this function

idryzhov replied:
It's just an old comment that I didn't touch.
It's just git treating this as a change.
I will remove it.

qlyoung reviewed (commented):
Can you fix the style? Other than that it looks good, although I'm not sure I understand the file pointer comment as I mentioned.

idryzhov commented:
Actually frrbot suggests to fix the code that I didn't even touch. Anyway, I'll update the style.

riw777 commented:
ci:rerun

idryzhov commented:
@qlyoung @riw777 I fixed the style, removed wrong comment and renamed one function as suggested by Pushpasis.
The PR is ready to be merged once CI passes. Thanks for all the reviews.
